src/pdu_loop/frame_element.rs

```diff
@@ -0,0 +1,494 @@
+        unsafe { self.inner.frame() }.command
+    }
+}
+
+pub struct ReceiveFrameFut<'sto> {
+    frame: Option<FrameBox<'sto>>,
+}
+
+impl<'sto> Future for ReceiveFrameFut<'sto> {
+    type Output = Result<ReceivedFrame<'sto>, Error>;
+
+    fn poll(
+        mut self: core::pin::Pin<&mut Self>,
+        cx: &mut core::task::Context<'_>,
+    ) -> Poll<Self::Output> {
+        let rxin = match self.frame.take() {
+            Some(r) => r,
+            None => return Poll::Ready(Err(PduError::InvalidFrameState.into())),
+        };
+
+        let swappy = unsafe {
+            FrameElement::swap_state(rxin.frame, FrameState::RxDone, FrameState::RxProcessing)
+        };
+
+        let was = match swappy {
+            Ok(_frame_element) => {
+                log::trace!("Frame future is ready");
+                return Poll::Ready(Ok(ReceivedFrame { inner: rxin }));
+            }
+            Err(e) => e,
+        };
+
+        match was {
+            FrameState::Sendable | FrameState::Sending => {
+                unsafe { rxin.replace_waker(cx.waker().clone()) };
+
+                self.frame = Some(rxin);
+
+                Poll::Pending
+            }
+            _ => Poll::Ready(Err(PduError::InvalidFrameState.into())),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ReceivedFrame<'sto> {
+    inner: FrameBox<'sto>,
+}
+
+impl<'sto> ReceivedFrame<'sto> {
+    fn working_counter(&self) -> u16 {
+        unsafe { self.inner.frame() }.working_counter
+    }
+
+    pub fn wkc(self, expected: u16, context: &'static str) -> Result<RxFrameDataBuf<'sto>, Error> {
+        let frame = self.frame();
+        let act_wc = frame.working_counter;
+
+        if act_wc == expected {
+            Ok(self.into_data_buf())
+        } else {
+            Err(Error::WorkingCounter {
+                expected,
+                received: act_wc,
+                context: Some(context),
+            })
+        }
+    }
+
+    /// Retrieve the frame's internal data without checking for working counter.
+    pub fn into_data(self) -> PduResponse<RxFrameDataBuf<'sto>> {
+        let wkc = self.working_counter();
+
+        (self.into_data_buf(), wkc)
+    }
+
+    fn frame(&self) -> &PduFrame {
+        unsafe { self.inner.frame() }
+    }
+
+    fn into_data_buf(self) -> RxFrameDataBuf<'sto> {
+        let len: usize = self.frame().flags.len().into();
+
+        // debug_assert!(len <= self.frame.buf_len);
+
+        let sptr = unsafe { FrameElement::buf_ptr(self.inner.frame) };
+        let eptr = unsafe { NonNull::new_unchecked(sptr.as_ptr().add(len)) };
+
+        RxFrameDataBuf {
+            _frame: self,
+            data_start: sptr,
+            data_end: eptr,
+        }
+    }
+}
+
+impl<'sto> Drop for ReceivedFrame<'sto> {
+    fn drop(&mut self) {
+        log::trace!("Drop frame element");
+
+        unsafe { FrameElement::set_state(self.inner.frame, FrameState::None) }
+    }
+}
+
+pub struct RxFrameDataBuf<'sto> {
```

Comment:
I'd like to understand how this solves lifetime issues in `send_coe_service`. I'm missing a lot of Rust-fu but also feel the answer is reasonably simple :D